File: ios/Shared/Sources/Core/Parsing/LmsHTMLTextFormatter.swift
import Foundation

public enum LmsHTMLTextFormatter {
    public static func plainText(from html: String?) -> String? {
        guard let html, html.isEmpty == false else {
            return nil
        }

        let withParagraphBreaks =
            html
            .replacingOccurrences(
                of: "(?i)<\\s*br\\s*/?>|</p>|</div>|</li>|</h[1-6]>",
                with: "\n",
                options: .regularExpression
            )
            .replacingOccurrences(of: "(?i)<li[^>]*>", with: "• ", options: .regularExpression)

        let withoutTags = withParagraphBreaks.replacingOccurrences(
            of: "<[^>]+>",
            with: " ",
            options: .regularExpression
        )

        let decodedEntities =
            withoutTags
            .replacingOccurrences(of: "&nbsp;", with: " ")
            .replacingOccurrences(of: "&amp;", with: "&")
            .replacingOccurrences(of: "&lt;", with: "<")
            .replacingOccurrences(of: "&gt;", with: ">")
            .replacingOccurrences(of: "&quot;", with: "\"")
            .replacingOccurrences(of: "&#39;", with: "'")

        let normalizedLines =
            decodedEntities
            .replacingOccurrences(of: "\r\n", with: "\n")
            .components(separatedBy: .newlines)
            .map {
                $0.replacingOccurrences(of: "\\s+", with: " ", options: .regularExpression)
                    .trimmingCharacters(in: .whitespacesAndNewlines)
            }
            .filter { $0.isEmpty == false }

        guard normalizedLines.isEmpty == false else {
            return nil
        }

        return normalizedLines.joined(separator: "\n\n")
    }
}

File: android/core/src/main/kotlin/dev/example/moodlenative/core/LmsHTMLTextFormatter.kt
package dev.example.moodlenative.core

object LmsHTMLTextFormatter {
    fun plainText(html: String?): String? {
        if (html.isNullOrEmpty()) {
            return null
        }

        val withParagraphBreaks = html
            .replace(Regex("(?i)<\\s*br\\s*/?>|</p>|</div>|</li>|</h[1-6]>"), "\n")
            .replace(Regex("(?i)<li[^>]*>"), "• ")
        val withoutTags = withParagraphBreaks.replace(Regex("<[^>]+>"), " ")
        val decodedEntities = withoutTags
            .replace("&nbsp;", " ")
            .replace("&amp;", "&")
            .replace("&lt;", "<")
            .replace("&gt;", ">")
            .replace("&quot;", "\"")
            .replace("&#39;", "'")

        val normalizedLines = decodedEntities
            .replace("\r\n", "\n")
            .lines()
            .map { it.replace(Regex("\\s+"), " ").trim() }
            .filter { it.isNotEmpty() }

        return normalizedLines.ifEmpty { null }?.joinToString(separator = "\n\n")
    }
}
